src/libs/ck-libs/NDMeshStreamer/NDMeshStreamer.h

```diff
@@ -631,15 +634,18 @@ storeMessage(int destinationPe, const Route& destinationRoute,
 }
 
 template <class dtype, class RouterType>
-inline void MeshStreamer<dtype, RouterType>::
-insertData(const DataItemHandle<dtype> *dataItemHandle, int destinationPe) {
-  // no data items should be submitted after all local contributors call done
-  // and staged completion has begun
+inline void MeshStreamer<dtype, RouterType>::createDetectors() {
+  // No data items should be submitted when staged completion has begun
   CkAssert(stagedCompletionStarted_ == false);
 
-  if (useCompletionDetection_) {
-    detectorLocalObj_->produce();
-  }
+  // Increment completion detection and quiescence detection
+  if (useCompletionDetection_) detectorLocalObj_->produce();
+  QdCreate(1);
+}
+
+template <class dtype, class RouterType>
+inline void MeshStreamer<dtype, RouterType>::
+insertData(const DataItemHandle<dtype> *dataItemHandle, int destinationPe) {
```

Comment:
Does this function need to call `if (useCompletionDetection_) detectorLocalObj_->produce();`? It seems like the old array code would call that once in its own `insertData` and then again in this one.

Reply:
So the old code had this weird way of creating completion and quiescence detections for group chares in `MeshStreamer::insertData` (which is eventually invoked by `GroupMeshStreamer` as it didn't have its own `insertData` function), whereas for chare arrays it was done in `ArrayMeshStreamer::insertData`. Since I added the `GroupMeshStreamer::insertData` in this PR (which creates the detectors like `ArrayMeshStreamer::insertData`), `MeshStreamer::insertData` shouldn't create the detectors anymore.

Reply:
I'm still a bit confused. In the old version, it looks like `ArrayMeshStreamer::insertData()` called `produce()`, then it calls `MeshStreamer::insertData()` (on line 1190), which called `produce()` again. This PR would reduce the number of calls to `produce()` but not to `consume()` for that path way, it seems, so they may not match anymore. Am I missing something here?

Reply:
Sorry by the old code I meant the version before Venkat's varTRAM patch (e.g. v6.9.0) that removed `GroupMeshStreamer`. Before the varTRAM patch there were two `MeshStreamer::insertData` functions (one that takes `const void *dataItemHandle` and another that takes `const dtype& dataItem`), and the second one called `produce()` and `QdCreate()` and then called the first `MeshStreamer::insertData` again. `ArrayMeshStreamer::insertData` called the first `MeshStreamer::insertData` directly (which doesn't call `produce()`, etc.) so only  one call to `produce()` and `QdCreate()` was made, which is correct. So Venkat's patch actually changed the `ArrayMeshStreamer::insertData` to call `produce()` two times, which is incorrect. FYI, `GroupMeshStreamer` called the second `MeshStreamer::insertData` before the varTRAM patch to call `produce()` and `QdCreate()`, and my PR effectively does the same by adding a separate GroupMeshStreamer::insertData that calls `createDetectors`.

Reply:
Okay, gotcha, thanks for the explanation.